internal/codeintel/codenav/mapped_index.go

```diff
@@ -0,0 +1,215 @@
+	// so we'd just be increasing resident memory
+	if document, ok := optDocument.Get(); ok {
+		return core.Some[MappedDocument](&mappedDocument{
+			gitTreeTranslator: i.gitTreeTranslator,
+			indexCommit:       i.upload.GetCommit(),
+			targetCommit:      i.targetCommit,
+			path:              path,
+			document: &lockedDocument{
+				inner:      document,
+				isMapped:   false,
+				mapErrored: nil,
+				lock:       sync.RWMutex{},
+			},
+			mapOnce: sync.Once{},
+		}), nil
+	} else {
+		return core.None[MappedDocument](), nil
+	}
+}
+
+type mappedDocument struct {
+	gitTreeTranslator GitTreeTranslator
+	indexCommit       api.CommitID
+	targetCommit      api.CommitID
+	path              core.RepoRelPath
+
+	document *lockedDocument
+	mapOnce  sync.Once
+}
+
+type lockedDocument struct {
+	inner      *scip.Document
+	isMapped   bool
+	mapErrored error
+	lock       sync.RWMutex
+}
+
+// Concurrency: Only call this while you're holding a read lock on the document
+func (d *mappedDocument) mapAllOccurrences(ctx context.Context) ([]*scip.Occurrence, error) {
+	newOccurrences := make([]*scip.Occurrence, 0)
+	for _, occ := range d.document.inner.Occurrences {
+		scipRange := scip.NewRangeUnchecked(occ.Range)
+		sharedRange := shared.TranslateRange(scipRange)
+		mappedRange, ok, err := d.gitTreeTranslator.GetTargetCommitRangeFromSourceRange(ctx, string(d.indexCommit), d.path.RawValue(), sharedRange, true)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			continue
+		}
+		newOccurrences = append(newOccurrences, &scip.Occurrence{
+			Range:                 mappedRange.ToSCIPRange().SCIPRange(),
+			Symbol:                occ.Symbol,
+			SymbolRoles:           occ.SymbolRoles,
+			OverrideDocumentation: occ.OverrideDocumentation,
+			SyntaxKind:            occ.SyntaxKind,
+			Diagnostics:           occ.Diagnostics,
+			EnclosingRange:        occ.EnclosingRange,
+		})
+	}
+	return newOccurrences, nil
+}
+
+func (d *mappedDocument) GetOccurrences(ctx context.Context) ([]*scip.Occurrence, error) {
+	// It's important we don't remap the occurrences twice
+	d.mapOnce.Do(func() {
+		d.document.lock.RLock()
+		newOccurrences, err := d.mapAllOccurrences(ctx)
+		d.document.lock.RUnlock()
+
+		d.document.lock.Lock()
+		defer d.document.lock.Unlock()
+		if err != nil {
+			d.document.mapErrored = err
+			return
+		}
+		d.document.inner.Occurrences = newOccurrences
+		d.document.isMapped = true
+	})
+
+	d.document.lock.RLock()
+	defer d.document.lock.RUnlock()
+	if d.document.mapErrored != nil {
+		return nil, d.document.mapErrored
+	}
+	return d.document.inner.Occurrences, nil
+}
+
+func (d *mappedDocument) GetOccurrencesAtRange(ctx context.Context, rg scip.Range) ([]*scip.Occurrence, error) {
+	d.document.lock.RLock()
+	defer d.document.lock.RUnlock()
+	if d.document.isMapped {
+		return FindOccurrencesWithEqualRange(d.document.inner.Occurrences, rg), nil
+	}
```

Comment:
This doesn't have data races AFAICT, but a bit too pessimistic.

```suggestion
	d.document.lock.RLock()
	occs := d.document.inner.Occurrences
	if d.document.isMapped {
	    d.document.lock.RUnlock()
		return FindOccurrencesWithEqualRange(occs, rg), nil
	}
	d.document.lock.RUnlock()
```

Let's reduce the size of the critical section. You can use `occs` later without extra synchronization instead of `d.document.inner.Occurrences`, since `occs` will refer to a local variable, and there isn't any code modifying the occurrences themselves.

If you don't like the duplication of `RUnlock`, you can introduce an immediately invoked function literal.